fhir-bucket/src/main/java/com/ibm/fhir/bucket/reindex/ClientDrivenReindexOperation.java

```diff
@@ -0,0 +1,359 @@
+/*
+ * (C) Copyright IBM Corp. 2021
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.ibm.fhir.bucket.reindex;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.LinkedBlockingDeque;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+
+import org.apache.http.HttpStatus;
+
+import com.ibm.fhir.bucket.client.FHIRBucketClient;
+import com.ibm.fhir.bucket.client.FHIRBucketClientUtil;
+import com.ibm.fhir.bucket.client.FhirServerResponse;
+import com.ibm.fhir.database.utils.api.DataAccessException;
+import com.ibm.fhir.model.resource.Parameters;
+import com.ibm.fhir.model.resource.Parameters.Builder;
+import com.ibm.fhir.model.resource.Parameters.Parameter;
+import com.ibm.fhir.model.resource.Resource;
+
+/**
+ * Drives the $reindex custom operation in parallel from the client side via use of the $retrieve-index operation.
+ * Processing continues until logical resource IDs indicate that no resources remain to be reindexed.
+ */
+public class ClientDrivenReindexOperation extends DriveReindexOperation {
+    private static final Logger logger = Logger.getLogger(ClientDrivenReindexOperation.class.getName());
+
+    private static final String COUNT_PARAM = "_count";
+    private static final String NOT_MODIFIED_AFTER_PARAM = "notModifiedAfter";
+    private static final String AFTER_LOGICAL_RESOURCE_ID_PARAM = "afterLogicalResourceId";
+    private static final String LOGICAL_RESOURCE_IDS_PARAM = "logicalResourceIds";
+    private static final int MAX_RETRIEVE_COUNT = 1000;
+    private static final int OFFER_TIMEOUT_IN_SEC = 10;
+    private static final String RETRIEVE_INDEX_URL = "$retrieve-index";
+    private static final String REINDEX_URL = "$reindex";
+
+    // Flags to indicate if we should be running
+    private volatile boolean running = true;
+    private volatile boolean active = false;
+    private volatile boolean doneRetrieving = false;
+
+    // FHIR client
+    private final FHIRBucketClient fhirClient;
+
+    // Maximum number of concurrent requests
+    private final int maxConcurrentRequests;
+
+    // Timestamp the reindex began
+    private final String reindexTimestamp;
+
+    // Maximum number of resources reindexed per thread
+    private final int maxResourceCount;
+
+    // Queue for holding logical resource IDs to reindex
+    private BlockingQueue<String> blockingQueue;
+
+    // Last logical resource ID found by monitor thread
+    private String lastLogicalResourceId;
+
+    // Monitor thread
+    private Thread monitorThread;
+
+    // Thread pool for processing requests
+    private final ExecutorService pool = Executors.newCachedThreadPool();
```

Comment:
Does it use `the number of cpu cores on the system - 1` in the pool?
Doe it relate to the passed `--reindex-concurrent-requests` value at all?

Reply:
Yes, the passed `--reindex-concurrent-requests` value determines the number of worker threads created in the thread pool in the monitorLoop method.  This was same logic as used in the ServerDrivenReindexOperation, so I followed that.